bolt-modules/boltlib/lib/puppet/functions/run_container.rb

```diff
@@ -0,0 +1,148 @@
+    cmd = %w[run]
+    cmd += Bolt::Util.format_env_vars_for_cli(options[:env_vars]) if options[:env_vars]
+    cmd += volumes if volumes
+    cmd += ports if ports
+    cmd << "--rm" if options[:rm]
+    cmd += %W[-w #{options[:workdir]}] if options[:workdir]
+    cmd << image
+    cmd += Shellwords.shellsplit(options[:cmd]) if options[:cmd]
+
+    executor.publish_event(type: :container_start, image: image)
+    out, err, status = Bolt::Util.exec_docker(cmd)
+    executor.publish_event(type: :container_finish, image: image, status: status.success?)
+
+    value = { 'stdout' => out, 'stderr' => err, 'exit_code' => status.exitstatus }
+
+    details = %w[file line].zip(Puppet::Pops::PuppetStack.top_of_stack).to_h.compact
+    unless status.exitstatus.zero?
+      details['exit_code'] = status.exitstatus
+      value = { '_error' => {
+        'kind' => 'puppetlabs.tasks/container-error',
+        'issue_code' => 'CONTAINER_ERROR',
+        'msg' => "Error running container '#{image}': #{err}",
+        'details' => details
+      } }
+      unless options[:catch_errors]
+        raise Bolt::ContainerFailure.new(value, image)
+      end
+    end
+
+    Bolt::ContainerResult.new(value: value, object: image)
+  end
+
+  def validate_options(options)
+    if (ev = options[:env_vars])
+      unless options[:env_vars].is_a?(Hash)
+        msg = "Option 'env_vars' must be a hash. Received #{ev} which is a #{ev.class}"
+        raise Bolt::ValidationError, msg
+      end
+
+      if (bad_keys = options[:env_vars].keys.reject { |k| k.is_a?(String) }).any?
+        raise Bolt::ValidationError,
+              "Keys for option 'env_vars' must be strings: #{bad_keys.map(&:inspect).join(', ')}"
+      end
+    end
+
+    if (volumes = options[:volumes])
+      unless options[:volumes].is_a?(Hash)
+        msg = "Option 'volumes' must be a hash. Received #{v_dest} which is a #{v_dest.class}"
+        raise Bolt::ValidationError, msg
+      end
+
+      if (bad_vs = volumes.reject { |k, v| k.is_a?(String) && v.is_a?(String) }).any?
+        raise Bolt::ValidationError,
+              "Option 'volumes' only accepts strings for keys and values. "\
+              "Received: #{bad_vs.map(&:inspect).join(', ')}"
+      end
+    end
+
+    if (cmd = options[:cmd]) && !options[:cmd].is_a?(String)
+      msg = "Option 'cmd' must be a string. Received #{cmd} which is a #{cmd.class}"
+      raise Bolt::ValidationError, msg
+    end
+
+    if (ports = options[:ports])
+      unless ports.is_a?(Hash)
+        msg = "Option 'ports' must be a hash. Received #{ports} which is a #{ports.class}"
+        raise Bolt::ValidationError, msg
+      end
+
+      if (bad_ps = ports.reject { |k, v| k.is_a?(Integer) && v.is_a?(Integer) }).any?
+        raise Bolt::ValidationError,
+              "Option 'ports' only accepts integers for keys and values. "\
+              "Received: #{bad_ps.map(&:inspect).join(', ')}"
```

Comment:
The validation messages switch between assigning the message to a var and passing a string literal to the initializer. We should try to be consistent just from a style standpoint.